frontend/src/wasm-communication/messages.ts

```diff
@@ -1188,13 +1190,46 @@ export type WidgetLayout = {
 	layout: LayoutGroup[];
 };
 
+export type WidgetDiffUpdate = {
+	layoutTarget: unknown;
+	diff: WidgetDiff[];
+};
+
+type WidgetDiff = { path: number[]; new: LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[] };
+
 export function defaultWidgetLayout(): WidgetLayout {
 	return {
 		layoutTarget: undefined,
 		layout: [],
 	};
 }
 
+// Updates a widget layout based on a list of updates, returning the new layout
+export function patchWidgetLayout(layout: WidgetLayout, updates: WidgetDiffUpdate): void {
+	layout.layoutTarget = updates.layoutTarget;
```

Comment:
If anything I think this should be an asset. I do not think there is a case where the update and the base do not have the same `layoutTarget`

Reply:
Well they don't on the first update.

Reply:
I wrote this before seeing that the system is diff only. I think a hybrid system is necessary so that diffs only patch existing layouts

Reply:
Currently on start we just send a diff with the full layout and a path of [] - which isn't too much overhead for a more consistent system.